libraries/rush-lib/src/logic/operations/ShardedPhaseOperationPlugin.ts

```diff
@@ -0,0 +1,162 @@
+
+  for (const operation of existingOperations) {
+    const { associatedPhase: phase, associatedProject: project } = operation;
+    const operationSettings: IOperationSettings | undefined = getOperationSettings(phase, project);
+    if (operationSettings?.sharding) {
+      shardableOperations.add(operation);
+    }
+  }
+  const getCustomParameterValuesForPhase: (phase: IPhase) => ReadonlyArray<string> =
+    getCustomParameterValuesByPhase();
+
+  for (const operation of shardableOperations) {
+    const { associatedPhase: phase, associatedProject: project } = operation;
+    const operationSettings: IOperationSettings | undefined = getOperationSettings(phase, project);
+    if (phase && project && operationSettings?.sharding && operationSettings.sharding.count > 1) {
+      existingOperations.delete(operation);
+
+      const collatorNode: Operation = new Operation({
+        phase,
+        project
+      });
+      existingOperations.add(collatorNode);
+      const collatorDisplayName: string = `${getDisplayName(phase, project)} - collate`;
+      const rawScript: string | undefined = project.packageJson.scripts?.[`${phase.name}:collate`];
+
+      if (rawScript) {
+        const collatorRunner: ShellOperationRunner = new ShellOperationRunner({
+          commandToRun: rawScript,
+          displayName: collatorDisplayName,
+          phase,
+          rushConfiguration,
+          rushProject: project,
+          operationSettings
+        });
+        collatorNode.runner = collatorRunner;
+      } else {
+        collatorNode.runner = new NullOperationRunner({
+          name: collatorDisplayName,
+          result: OperationStatus.NoOp,
+          silent: phase.missingScriptBehavior === 'silent'
+        });
+      }
+      for (const dependent of operation.consumers) {
+        dependent.addDependency(collatorNode);
+        dependent.deleteDependency(operation);
+      }
+
+      const shards: number = operationSettings.sharding.count;
+      for (const shard of Array.from({ length: shards }, (element, index) => index + 1)) {
+        let customParameters: readonly string[] = getCustomParameterValuesForPhase(phase);
+
+        // Add the shard argument to the custom parameters, replacing any templated values.
+        const shardArgumentFormat: string =
+          operationSettings.sharding.shardArgumentFormat ?? '--shard={shardIndex}/{shardCount}';
+        const outputDirectory: string = `.rush/shards/${shard}`;
+        const shardArgument: string = shardArgumentFormat
+          .replace('{shardIndex}', shard.toString())
+          .replace('{shardCount}', shards.toString());
+        const outputDirectoryArgument: string = `--shard-output-directory="${outputDirectory}"`;
+        customParameters = [...customParameters, shardArgument, outputDirectoryArgument];
+
+        const commandToRun: string | undefined = getScriptToRun(
+          project,
+          phase.name,
```

Comment:
I thought the last state we left the design in during the call was that the collator node was `_phase:foo` and the shards were `_phase:foo:shard`? That was `_phase:foo` is the node that exits the original phase, and `_phase:foo:shard` is the one that gets special arguments that aren't part of the default phase definition.

Reply:
See my comment [here](https://github.com/microsoft/rushstack/pull/4652#discussion_r1581247792), I think the `{phase.name}:collate` format makes it a little clearer what is happening and requires fewer changes for people adopting this script, ie "`_phase:test` is still testing but I need a collate command to collate coverage"

Reply:
Thinking on this some more, I'll adjust to the `:shard` model. The existing config like output folders and phase name should consistently refer to something, where currently they don't because shard is the primary operation, so `outputFolders` is used for the collation but `operationName` and config from `command-line.json` refers to the sharding operation.

Reply:
Updated in [058eee7](https://github.com/microsoft/rushstack/pull/4652/commits/058eee73433bbcb406f0d9798642b575f0c3904b)